File: studio/AuraStudio/Sources/AuraStudio/Services/TMDBClient.swift
import Foundation

/// Cliente de The Movie Database (ST-022) -- API v3, key propia
/// (gratuita, `APIKeyService.tmdb`, en el Keychain). Cumple dos papeles
/// para los pósters de video:
/// 1. **Resolvedor de identificadores**: fanart.tv no busca por título;
///    indexa películas por ID de TMDB/IMDb y series por ID de TheTVDB.
///    TMDB resuelve título → ID de película, y título → ID de serie →
///    `external_ids.tvdb_id`.
/// 2. **Póster de respaldo**: TMDB trae su propio `poster_path`
///    (`image.tmdb.org`), que se usa cuando fanart.tv no tiene el
///    título (o no hay key de fanart.tv).
/// Sin key no toca la red (nil), como los demás clientes opcionales.
struct TMDBClient {
    struct Movie: Decodable, Equatable {
        let id: Int
        let title: String
        let releaseDate: String?
        let posterPath: String?

        enum CodingKeys: String, CodingKey {
            case id, title
            case releaseDate = "release_date"
            case posterPath = "poster_path"
        }

        var year: String? { releaseDate.map { String($0.prefix(4)) }.flatMap { $0.isEmpty ? nil : $0 } }
    }

    struct TVShow: Decodable, Equatable {
        let id: Int
        let name: String
        let firstAirDate: String?
        let posterPath: String?

        enum CodingKeys: String, CodingKey {
            case id, name
            case firstAirDate = "first_air_date"
            case posterPath = "poster_path"
        }

        var year: String? { firstAirDate.map { String($0.prefix(4)) }.flatMap { $0.isEmpty ? nil : $0 } }
    }

    private struct SearchResponse<T: Decodable>: Decodable {
        let results: [T]
    }

    private struct ExternalIDs: Decodable {
        let tvdbID: Int?
        let imdbID: String?

        enum CodingKeys: String, CodingKey {
            case tvdbID = "tvdb_id"
            case imdbID = "imdb_id"
        }
    }

    private let session: URLSession
    private let baseURL: URL
    private let imageBaseURL: URL
    private let apiKeyProvider: @Sendable () -> String?
    /// Idioma de títulos/pósters: los pósters "es-MX" existen para casi
    /// todo lo popular y caen a los originales cuando no.
    private let language: String

    init(session: URLSession = .shared,
         baseURL: URL = URL(string: "https://api.themoviedb.org/3")!,
         imageBaseURL: URL = URL(string: "https://image.tmdb.org/t/p/w780")!,
         language: String = "es-MX",
         apiKeyProvider: @escaping @Sendable () -> String? = { APIKeyStore.load(for: .tmdb) }) {
        self.session = session
        self.baseURL = baseURL
        self.imageBaseURL = imageBaseURL
        self.language = language
        self.apiKeyProvider = apiKeyProvider
    }

    var hasAPIKey: Bool { apiKeyProvider() != nil }

    /// Película por título (y año si se conoce -- afina mucho con
    /// remakes homónimos). Devuelve el primer resultado: TMDB ya ordena
    /// por relevancia y popularidad.
    func searchMovie(title: String, year: String? = nil) async throws -> Movie? {
        var items = [URLQueryItem(name: "query", value: title), URLQueryItem(name: "include_adult", value: "false")]
        if let year, !year.isEmpty { items.append(URLQueryItem(name: "year", value: year)) }
        let response: SearchResponse<Movie>? = try await get("search/movie", query: items)
        return response?.results.first
    }

    func searchTV(name: String, year: String? = nil) async throws -> TVShow? {
        var items = [URLQueryItem(name: "query", value: name), URLQueryItem(name: "include_adult", value: "false")]
        if let year, !year.isEmpty { items.append(URLQueryItem(name: "first_air_date_year", value: year)) }
        let response: SearchResponse<TVShow>? = try await get("search/tv", query: items)
        return response?.results.first
    }

    /// ID de TheTVDB de una serie -- lo único que fanart.tv acepta para TV.
    func tvdbID(forTVShow id: Int) async throws -> Int? {
        let ids: ExternalIDs? = try await get("tv/\(id)/external_ids", query: [])
        return ids?.tvdbID
    }

    /// Descarga el póster de TMDB (`poster_path` relativo, p. ej.
    /// `/abc.jpg`) a 780 px de ancho -- de sobra para los ≤640 px que
    /// admite el iPod, sin bajar el original de 2000 px.
    func downloadPoster(path: String?) async throws -> Data? {
        guard let path, !path.isEmpty else { return nil }
        let url = imageBaseURL.appendingPathComponent(path)
        let (data, response) = try await session.data(from: url)
        try MusicBrainzClient.validate(response)
        return data
    }

    private func get<T: Decodable>(_ path: String, query: [URLQueryItem]) async throws -> T? {
        guard let apiKey = apiKeyProvider() else { return nil }
        var components = URLComponents(url: baseURL.appendingPathComponent(path), resolvingAgainstBaseURL: false)!
        components.queryItems = query + [
            URLQueryItem(name: "language", value: language),
            URLQueryItem(name: "api_key", value: apiKey),
        ]
        var request = URLRequest(url: components.url!)
        request.setValue(MusicBrainzClient.userAgent, forHTTPHeaderField: "User-Agent")
        let (data, response) = try await session.data(for: request)
        if let http = response as? HTTPURLResponse, http.statusCode == 404 { return nil }
        try MusicBrainzClient.validate(response)
        return try JSONDecoder().decode(T.self, from: data)
    }
}

/// Saca título, año y (si es serie) nombre de la serie del nombre de
/// archivo o del título que ya tenga el video (ST-022). Los nombres
/// reales vienen como `The.Matrix.1999.1080p.BluRay.x264.mkv` o
/// `Breaking Bad - S01E02 - Cat's in the Bag.mp4`; sin limpiar eso, la
/// búsqueda en TMDB no encuentra nada.
enum VideoTitleParser {
    struct Parsed: Equatable {
        var title: String
        var year: String?
        /// Nombre de la serie cuando el archivo trae `SxxEyy` / `1x02`.
        var seriesName: String?
        var season: Int?
        var episode: Int?

        var isEpisode: Bool { seriesName != nil }
    }

    private static let noiseTokens: Set<String> = [
        "1080p", "720p", "480p", "2160p", "4k", "uhd", "hdr", "hdr10", "dv", "x264", "x265", "h264", "h265", "hevc",
        "avc", "aac", "ac3", "dts", "bluray", "bdrip", "brrip", "webrip", "web-dl", "webdl", "web", "hdtv", "dvdrip",
        "dvd", "remux", "proper", "repack", "extended", "unrated", "remastered", "multi", "dual", "latino", "castellano",
        "subs", "sub", "esp", "eng", "spa", "lat", "amzn", "nf", "dsnp", "hmax", "atvp", "yify", "yts", "rarbg", "10bit",
        "5.1", "7.1", "ddp5.1", "dd5.1", "atmos", "imax", "hq", "xvid", "divx", "mkv", "mp4", "avi",
    ]

    static func parse(_ raw: String) -> Parsed {
        var text = raw
            .replacingOccurrences(of: "_", with: " ")
            .replacingOccurrences(of: ".", with: " ")
        // Quitar etiquetas entre corchetes ([1080p], [Latino], [grupo]).
        text = text.replacingOccurrences(of: #"\[[^\]]*\]"#, with: " ", options: .regularExpression)

        var parsed = Parsed(title: text)

        // Serie: "S01E02", "s1e2", "1x02".
        if let match = text.range(of: #"(?i)\bS(\d{1,2})\s?E(\d{1,3})\b|\b(\d{1,2})x(\d{2,3})\b"#, options: .regularExpression) {
            let marker = String(text[match])
            let numbers = marker.components(separatedBy: CharacterSet.decimalDigits.inverted).filter { !$0.isEmpty }
            if numbers.count >= 2 {
                parsed.season = Int(numbers[0])
                parsed.episode = Int(numbers[1])
            }
            let before = String(text[..<match.lowerBound])
            let seriesName = cleanTitle(before)
            parsed.seriesName = seriesName.isEmpty ? nil : seriesName
            text = before
        }

        // Año entre 1900 y 2099, con o sin paréntesis; el ÚLTIMO que
        // aparezca (un título como "1917 (2019)" tiene dos).
        if let regex = try? NSRegularExpression(pattern: #"\(?\b(19\d{2}|20\d{2})\b\)?"#) {
            let range = NSRange(text.startIndex..., in: text)
            let matches = regex.matches(in: text, range: range)
            if let last = matches.last, let yearRange = Range(last.range(at: 1), in: text), let fullRange = Range(last.range, in: text) {
                let year = String(text[yearRange])
                // Si el "año" es lo único que hay (película que se llama
                // "2012"), no se lo quites al título.
                let remainder = cleanTitle(String(text[..<fullRange.lowerBound]))
                if !remainder.isEmpty {
                    parsed.year = year
                    text = String(text[..<fullRange.lowerBound])
                }
            }
        }

        parsed.title = cleanTitle(text)
        if parsed.title.isEmpty { parsed.title = cleanTitle(raw) }
        return parsed
    }

    /// Quita tokens de ruido (calidad, códec, grupo), guiones sueltos y
    /// espacios repetidos.
    static func cleanTitle(_ text: String) -> String {
        let tokens = text
            .replacingOccurrences(of: #"\s+"#, with: " ", options: .regularExpression)
            .split(separator: " ")
            .map(String.init)
        var kept: [String] = []
        for token in tokens {
            let lower = token.lowercased().trimmingCharacters(in: CharacterSet(charactersIn: "-–()"))
            if lower.isEmpty { continue }
            if noiseTokens.contains(lower) { break } // lo que sigue al primer token de ruido es más ruido
            kept.append(token.trimmingCharacters(in: CharacterSet(charactersIn: "()")))
        }
        return kept.joined(separator: " ")
            .trimmingCharacters(in: CharacterSet(charactersIn: " -–"))
    }
}

/// Orquesta la búsqueda de póster para un video (ST-022): TMDB resuelve
/// el título; fanart.tv aporta el póster curado si tiene el título;
/// si no, se usa el póster de TMDB. Mejor esfuerzo, sin tirar.
struct VideoArtworkResolver {
    var tmdb: TMDBClient = TMDBClient()
    var fanart: FanartTVClient = FanartTVClient()
    var hasFanartKey: @Sendable () -> Bool = { APIKeyStore.hasKey(for: .fanartTV) }

    enum Kind { case movie, series, unknown }
    enum Source: Equatable { case fanartTV, tmdb }

    struct Result: Equatable {
        let data: Data
        let source: Source
        let matchedTitle: String
        let year: String?
    }

    enum Failure: Error, Equatable {
        case missingTMDBKey
        case noMatch
        case noPoster
    }

    /// `kind` viene de la categoría del video en Studio (Películas /
    /// Series / Videos): con `unknown` se prueba película y después
    /// serie -- o al revés si el nombre trae `SxxEyy`.
    func resolve(rawTitle: String, kind: Kind) async -> Swift.Result<Result, Failure> {
        guard tmdb.hasAPIKey else { return .failure(.missingTMDBKey) }
        let parsed = VideoTitleParser.parse(rawTitle)
        let order: [Kind]
        switch kind {
        case .movie: order = [.movie]
        case .series: order = [.series]
        case .unknown: order = parsed.isEpisode ? [.series, .movie] : [.movie, .series]
        }
        var sawMatch = false
        for candidate in order {
            switch candidate {
            case .movie:
                guard let movie = try? await tmdb.searchMovie(title: parsed.title, year: parsed.year) else { continue }
                sawMatch = true
                if hasFanartKey(), let data = try? await fanart.fetchMoviePoster(tmdbOrIMDbID: String(movie.id)), !data.isEmpty {
                    return .success(Result(data: data, source: .fanartTV, matchedTitle: movie.title, year: movie.year))
                }
                if let data = try? await tmdb.downloadPoster(path: movie.posterPath), !data.isEmpty {
                    return .success(Result(data: data, source: .tmdb, matchedTitle: movie.title, year: movie.year))
                }
            case .series:
                let name = parsed.seriesName ?? parsed.title
                guard let show = try? await tmdb.searchTV(name: name, year: parsed.isEpisode ? nil : parsed.year) else { continue }
                sawMatch = true
                if hasFanartKey(),
                   let tvdb = try? await tmdb.tvdbID(forTVShow: show.id),
                   let data = try? await fanart.fetchTVPoster(tvdbID: String(tvdb)), !data.isEmpty {
                    return .success(Result(data: data, source: .fanartTV, matchedTitle: show.name, year: show.year))
                }
                if let data = try? await tmdb.downloadPoster(path: show.posterPath), !data.isEmpty {
                    return .success(Result(data: data, source: .tmdb, matchedTitle: show.name, year: show.year))
                }
            case .unknown:
                continue
            }
        }
        return .failure(sawMatch ? .noPoster : .noMatch)
    }
}

File: studio/windows/AuraStudio.Core/Networking/TMDBClient.cs
using System.Globalization;
using System.Net;
using System.Net.Http;
using System.Text.Json;
using System.Text.Json.Serialization;
using AuraStudio.Core.Resources;

namespace AuraStudio.Core.Networking;

/// <summary>
/// Cliente de The Movie Database (ST-033) -- API v3, key propia
/// (gratuita). Cumple dos papeles para los pósters de video:
/// 1. <b>Resolvedor de identificadores</b>: fanart.tv no busca por título;
///    indexa películas por ID de TMDB/IMDb y series por ID de TheTVDB.
///    TMDB resuelve título → ID de película, y título → ID de serie →
///    <c>external_ids.tvdb_id</c>.
/// 2. <b>Póster de respaldo</b>: TMDB trae su propio <c>poster_path</c>
///    (<c>image.tmdb.org</c>), que se usa cuando fanart.tv no tiene el
///    título (o no hay key de fanart.tv).
/// Sin key no toca la red (null), como los demás clientes opcionales.
/// </summary>
public sealed class TMDBClient
{
    /// <summary>Película encontrada por título (y año).</summary>
    public sealed class Movie
    {
        [JsonPropertyName("id")] public int Id { get; init; }
        [JsonPropertyName("title")] public string Title { get; init; } = "";
        [JsonPropertyName("release_date")] public string? ReleaseDate { get; init; }
        [JsonPropertyName("poster_path")] public string? PosterPath { get; init; }

        /// <summary>Año de estreno (extraído de release_date).</summary>
        public string? Year => ReleaseDate?.Length >= 4 ? ReleaseDate[..4] : null;
    }

    /// <summary>Serie de TV encontrada por nombre.</summary>
    public sealed class TVShow
    {
        [JsonPropertyName("id")] public int Id { get; init; }
        [JsonPropertyName("name")] public string Name { get; init; } = "";
        [JsonPropertyName("first_air_date")] public string? FirstAirDate { get; init; }
        [JsonPropertyName("poster_path")] public string? PosterPath { get; init; }

        /// <summary>Año de premiere (extraído de first_air_date).</summary>
        public string? Year => FirstAirDate?.Length >= 4 ? FirstAirDate[..4] : null;
    }

    /// <summary>IDs externos (TheTVDB, IMDb) de una serie.</summary>
    private sealed class ExternalIDs
    {
        [JsonPropertyName("tvdb_id")] public int? TvdbID { get; init; }
        [JsonPropertyName("imdb_id")] public string? ImdbID { get; init; }
    }

    private sealed class SearchResponse<T>
    {
        [JsonPropertyName("results")] public List<T> Results { get; init; } = new();
    }

    private readonly HttpClient _http;
    private readonly string _baseURL;
    private readonly string _imageBaseURL;
    private readonly IApiKeyStore _apiKeyStore;
    private readonly string? _language;

    /// <param name="language">
    /// Idioma en el que se le piden los datos a TMDB. Por omisión, el de la
    /// interfaz.
    ///
    /// <para>Estaba fijo en <c>"es-MX"</c>, y con la app en un solo idioma eso
    /// era correcto. Con seis deja de serlo: de acá salen los títulos y las
    /// sinopsis de películas y series, así que la app en japonés le mostraba al
    /// usuario texto en español recién traído de la red. No es una cadena de
    /// interfaz —es un parámetro de una consulta—, y por eso ninguna prueba de
    /// extracción lo iba a ver nunca (ST-247, B7c).</para>
    /// </param>
    public TMDBClient(
        HttpClient? http = null,
        IApiKeyStore? apiKeyStore = null,
        string baseURL = "https://api.themoviedb.org/3",
        string imageBaseURL = "https://image.tmdb.org/t/p/w780",
        string? language = null)
    {
        _http = http ?? new HttpClient();
        _baseURL = baseURL;
        _imageBaseURL = imageBaseURL;
        _apiKeyStore = apiKeyStore ?? new SimpleApiKeyStore();
        _language = language;
    }

    /// <summary>
    /// El idioma que viaja en la consulta. Se resuelve al consultar y no al
    /// construir el cliente, para que valga la cultura del momento.
    ///
    /// <para>La cultura invariante no tiene nombre y dejaría
    /// <c>&amp;language=</c> vacío: ahí la interfaz se ve en español, que es la
    /// cultura neutra, y eso es lo que se pide.</para>
    /// </summary>
    private string Language
    {
        get
        {
            if (_language is { Length: > 0 }) return _language;

            string name = CultureInfo.CurrentUICulture.Name;
            return name.Length > 0 ? name : AppLanguages.NeutralCulture;
        }
    }

    /// <summary>True si hay clave de API configurada.</summary>
    public bool HasAPIKey => _apiKeyStore.Load("tmdb") != null;

    /// <summary>Película por título (y año si se conoce). Devuelve el primer resultado.</summary>
    public async Task<Movie?> SearchMovieAsync(string title, string? year = null, CancellationToken ct = default)
    {
        var query = BuildQuery(title, year);
        var response = await GetAsync<SearchResponse<Movie>>("search/movie", query, ct).ConfigureAwait(false);
        return response?.Results.FirstOrDefault();
    }

    /// <summary>Serie de TV por nombre (y año si se conoce).</summary>
    public async Task<TVShow?> SearchTVAsync(string name, string? year = null, CancellationToken ct = default)
    {
        var query = BuildQuery(name, year);
        // TMDB usa first_air_date_year para series, no year
        var q = new List<KeyValuePair<string, string>>();
        foreach (var (k, v) in query) q.Add(new(k, v));
        q.Add(new("first_air_date_year", year ?? ""));
        q.RemoveAll(x => string.IsNullOrEmpty(x.Value));
        var response = await GetAsync<SearchResponse<TVShow>>("search/tv", q, ct).ConfigureAwait(false);
        return response?.Results.FirstOrDefault();
    }

    /// <summary>ID de TheTVDB de una serie — lo único que fanart.tv acepta para TV.</summary>
    public async Task<int?> GetTvdbIDAsync(int tvShowId, CancellationToken ct = default)
    {
        var ids = await GetAsync<ExternalIDs>($"tv/{tvShowId}/external_ids", [], ct).ConfigureAwait(false);
        return ids?.TvdbID;
    }

    /// <summary>
    /// Descarga el póster de TMDB (<c>poster_path</c> relativo, p. ej.
    /// <c>/abc.jpg</c>) a 780 px de ancho — de sobra para los ≤640 px que
    /// admite el iPod, sin bajar el original de 2000 px.
    /// </summary>
    public async Task<byte[]?> DownloadPosterAsync(string? path, CancellationToken ct = default)
    {
        if (string.IsNullOrEmpty(path)) return null;
        var url = $"{_imageBaseURL}{path}";
        using var request = new HttpRequestMessage(HttpMethod.Get, url);
        using var response = await _http.SendAsync(request, ct).ConfigureAwait(false);
        if (!response.IsSuccessStatusCode) return null;
        return await response.Content.ReadAsByteArrayAsync(ct).ConfigureAwait(false);
    }

    private List<KeyValuePair<string, string>> BuildQuery(string title, string? year)
    {
        var q = new List<KeyValuePair<string, string>>
        {
            new("query", title),
            new("include_adult", "false"),
        };
        if (!string.IsNullOrEmpty(year)) q.Add(new("year", year));
        return q;
    }

    private async Task<T?> GetAsync<T>(string path, List<KeyValuePair<string, string>> query, CancellationToken ct)
        where T : class
    {
        var apiKey = _apiKeyStore.Load("tmdb");
        if (apiKey == null) return null;

        var qs = string.Join("&", query
            .Where(x => !string.IsNullOrEmpty(x.Value))
            .Select(x => $"{Uri.EscapeDataString(x.Key)}={Uri.EscapeDataString(x.Value)}"));
        qs += $"&language={Uri.EscapeDataString(Language)}&api_key={Uri.EscapeDataString(apiKey)}";

        var url = $"{_baseURL}/{path}?{qs}";
        using var request = new HttpRequestMessage(HttpMethod.Get, url);
        request.Headers.TryAddWithoutValidation("User-Agent", HttpUserAgent.Value);

        using var response = await _http.SendAsync(request, ct).ConfigureAwait(false);
        if (response.StatusCode == HttpStatusCode.NotFound) return null;
        HttpResponseGuard.EnsureSuccess(response);

        var json = await response.Content.ReadAsStringAsync(ct).ConfigureAwait(false);
        return JsonSerializer.Deserialize<T>(json, new JsonSerializerOptions { PropertyNameCaseInsensitive = true });
    }

    /// <summary>
    /// Implementación simple que devuelve null (para cuando no se inyecta
    /// <c>IApiKeyStore</c> real en Core — el DefaultApiKeyStore se usa en
    /// la implementación real de Windows).
    /// </summary>
    private sealed class SimpleApiKeyStore : IApiKeyStore
    {
        public string? Load(string service) => null;
    }
}